sqlparser create table support

demmer opened the pull request:
Add support for parsing CREATE TABLE statements, including the various data types and column options.

This is the first in a series of PRs to support vtexplain (#3062).

bbeaudreault commented:
Do you have an idea of what constructs are unsupported now that we are parsing the full query? At hubspot we give developers pretty free reign so I imagine we have a broad range of create table syntax usages. The benefit of the current approach is those were all guaranteed to work. MySQL does the validation and vttablet propogates the error to the user. 

 In my experience In vitess we only add new more specific query parsing when we need to actually use those parts of the AST in other places in vitess, for routing for example. Are you planning on using these?

bbeaudreault commented:
Sorry I missed commenting on this in the issue before you started. I'm not sure what risk there is passing the DDLs straight through. If they are bad, MySQL will return an error. I do see risk in reformatting the queries, in the case of bugs and syntax errors which either cause DDLs to fail or cause schema to be subtly different from what one intended.

demmer commented:
As shown by the Travis results the parser isn't quite at 100% compatibility...

That said I think it is pretty complete. It parses the schema from Slack's largest pool of mysql shards and if issues come up (like the test failures) I should be able to easily fix them up once I'm back at a keyboard with a bit of time. Would you be willing to share hubspot's current schemas? If so we can certainly ensure compatibility for at least what you have now before merging this.

More fundamentally this change doesn't change how vitess actually handles the DDL statements. They are parsed differently but the original sql statements are still passed through unmodified to mysql, so I'm not worried about subtle bugs due to mismatches in the generated version as compared to the original.

Ideally we could come up with some way that the parser could run in a mode where it tries to parse the full statement but still succeeds even if it hits a construct that there isn't explicit support for like it used to. I presume this must be possible somehow but I don't have enough familiarity with goyacc internals to know how to do this.

bbeaudreault commented:
We have hundreds (or close to 1000+) of different schemas across a few hundred databases, owned by 100+ autonomous teams. I could work to compile a schema only dump of all of them, but I am concerned that it may not be accurate enough.  Doing `show create table` shows how to create the particular table in one canonical way, but if there's anything I've learned from my time in the sql parser it's that in MySQL there are typically numerous ways to do the same thing. So that dump may not represent how the tables were originally created, with subtle differences like ordering of keywords, optional keywords, default values, etc.

Is there a reason this is necessary? Why does vitess need to validate them, if mysql will just be the real validation when they are executed against it? Are you planning to use the AST in some way?

sougou commented:
I'm actually thinking that it should be possible to gather the minimal info necessary for a pass-through if parsing fails, like the table names involved in the statement. This will allow us to create the old sqlparser.DDL object which will work as before.

I still need to find time to figure out how.

sougou commented on go/vt/sqlparser/sql.y:
```diff
@@ -354,10 +391,449 @@ set_statement:
     $$ = &Set{Comments: Comments($2), Exprs: $3}
   }
 
+column_type:
```
It's better to move these at the end. The lose coding standard I've followed is to have to top level statements on top and the lower level ones at the bottom.

sougou commented on go/vt/sqlparser/sql.y:
```diff
@@ -354,10 +391,449 @@ set_statement:
+
 create_statement:
-  CREATE TABLE not_exists_opt table_name ddl_force_eof
+  CREATE TABLE not_exists_opt table_name table_spec
```
Can you try this construct:
```
func setDDL(yylex interface{}, ddl *DDL) {
  yylex.(*Tokenizer).ddl = ddl  // add ddl as member of Tokenizer
}
...


create_statement:
  create_prefix tablet_spec
  {
    $0.TableSpec = $2
  }

create_prefix:
  CREATE TABLE not_exists_opt table_name
  {
    $$ = &DDL{...}
    setDDL(yylex, $$)
  }
```

If you did this, we should still be able to return a successful DDL object if the `CREATE` fails to parse.

demmer commented:
As per @sougou's suggestion, I've modified the parser so it now preserves the previous tolerant treatment of CREATE TABLE statements by stashing the partially parsed DDL object in the Tokenizer and returning it even if the rest of the statement parsing fails.

@bbeaudreault I think this should address the core of your (and my) concern that incomplete support in the grammar could cause a regression for ApplySchema or other DDL operations.

I also modified the support for `ALTER TABLE` since there are now additional keywords in the grammar that needed to be added to the `non_rename_keyword` rule. With both these changes it passed at least one run of Travis.

I also cleaned up the CodeClimate complaints and fixed the couple other stylistic suggestions that @sougou had.

Finally, I added all the new tokens to what I think are the appropriate lists of either `reserved_keyword` or `non_reserved_keyword` baed on the mysql reference documents. @bbeaudreault can you please take a look and let me know if you think this looks correct?

bbeaudreault commented:
Thanks @demmer and @sougou for that backwards compatibility. One request on that: any chance we can increment a counter and/or do a warn when it occurs? Then we'll have the data in the future to see whether it's still needed.

In terms of the keywords, the usual approach is to try adding them to non_reserved_keywords first. If a goyacc error occurs, then move to reserved_keywords.  So I'd go through that process with each new keyword, preferring non_reserved over reserved.

The non_reserved_keywords are allowed in many more places, i.e. as column names. Vitess backticks them to ensure they get treated kindly by mysql. Despite the similar names, these 2 lists probably wont mirror MySQL much -- reason being they serve the same purpose for each, but the vitess grammar is very different from the mysql one.

Let me know if that makes sense.

demmer commented:
@bbeaudreault Thanks that does make sense! I updated the PR so now the newly-added keywords are in the non-reserved set.

I also added the warning log as per your suggestion.

sougou commented:
This looks good to me.
@bbeaudreault want to give it a once over?

bbeaudreault commented on go/vt/sqlparser/sql.y:
```diff
@@ -1288,39 +1778,70 @@ match_option:
 
 convert_type:
-  charset
+  BINARY length_opt
```
trying to figure out what happened here. it looks like convert_type just totally changed?

demmer replied:
Yeah... I made this note in the very first commit on this PR (thanks for reminding me).

> This change required modifying our handling of CONVERT statements since there are now several new keywords and we actually have grammar rules for the data types. As such the new grammar is more restrictive than it was before, although it passes all tests and should cover all the cases listed in the mysql 5.7 reference docs.

bbeaudreault replied:
Ok. I'll keep an eye out for syntax errors then, sound good

bbeaudreault commented:
LGTM

sougou commented:
LGTM
